Add validation of Unicode in bytes functions

TonyLianLong opened the pull request:
Add validation of Unicode in bytes functions and will raise
UnicodeDecodeError if the string is not a valid UTF-8 string.
See https://github.com/micropython/micropython/issues/2734

pfalcon reviewed (changes requested):
Please follow guidelines in https://github.com/micropython/micropython/blob/master/CONTRIBUTING.md . 

This functionality should be optional.

Please provide numbers about code growth and try to minimize it (e.g. mp_raise_UnicodeDecodeError() is not needed).

TonyLianLong committed:
py/objstr: Add validation of Unicode in bytes functions

Add validation of Unicode in bytes functions and will raise
UnicodeDecodeError if the string is not a valid UTF-8 string.

TonyLianLong commented:
I've changed it to raise only a ValueError and remove the UnicodeDecodeError.
`The size of micropython:
__TEXT	__DATA	__OBJC	others	dec	hex
331776	57344	0	4294998584	4295387704	100066a38	`

`tll:unix tll$ ./micropython -v
MicroPython v1.9.1-47-gf69e3af6 on 2017-06-24; darwin version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> bytes.decode(b"\xa1\x80", 'utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Not a valid UTF-8 string
>>> 
mem: total=4486, current=2472, peak=3825
stack: 304 out of 80000
GC: total: 2072832, used: 3840, free: 2068992
 No. of 1-blocks: 63, 2-blocks: 6, max blk sz: 41, max free sz: 64654
tll:unix tll$ `

dpgeorge commented on py/objstr.c:
```diff
@@ -132,6 +132,32 @@ STATIC void str_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t
     }
 }
 
+int utf8_chk(uint8_t *p, int len) {
```
Probably cleaner to put this in py/unicode.c.

dpgeorge commented on py/objstr.c:
```diff
@@ -161,6 +187,7 @@ mp_obj_t mp_obj_str_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_
                 if (str_hash == 0) {
                     str_hash = qstr_compute_hash(str_data, str_len);
                 }
+                if(!utf8_chk((uint8_t *)str_data, str_len)) mp_raise_ValueError("Not a valid UTF-8 string");
```
Since the UnicodeError exception object is provided by uPy it would be good to use that here, like `mp_raise_msg(&mp_type_UnicodeError, "invalid utf8 string")`.  Also please not coding style, and have a single suite with braces.

dpgeorge reviewed (commented):
Thanks for the contribution.  Are you able to also write some tests to test the new functionality?

TonyLianLong committed:
py and tests: Add a test and move the checking function into py/unicode.c

tests: add a test of invalid UTF-8 string
py: put utf_chk into unicode.c and change the code style in py/objstr.c

TonyLianLong commented:
I've added a test and change the error into UnicodeError.
Now:
```unix tll$ ./micropython
MicroPython v1.9.1-47-gf69e3af6-dirty on 2017-06-26; darwin version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> bytes.decode(b"\xa1\x80", 'utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeError: invalid utf8 string
>>> 
```

TonyLianLong commented:
I don't know why Travis CI test failed since I didn't edit any other code. Can anyone help me?

TonyLianLong committed:
py/objstr.c : do the check only when STR_UNICODE is specified

Do the check only when MICROPY_PY_BUILTINS_STR_UNICODE is specified
otherwise it will cause an error on minimal building

TonyLianLong committed:
py/unicode.c: Add STR_UNICODE check

To prevent unnecessary increase in core size, I add this check in
py/unicode.c

TonyLianLong committed:
py/unicode.h: Add the check in unicode.h

Also add the MICROPY_PY_BUILTINS_STR_UNICODE check in unicode.h

TonyLianLong committed:
Merge remote-tracking branch 'upstream/master'

TonyLianLong commented:
Looks like the core is increasing by 4 bytes which causes the validation to fail but I don't know where the 4 bytes come from. Can anyone help me?

dpgeorge commented:
> Looks like the core is increasing by 4 bytes which causes the validation to fail but I don't know where the 4 bytes come from.

It's probably because your patch is not against master but from a point a week or so ago.  And in the meantime code size decreased for the minimal port.  So you don't really need to worry about it, but to fix it you can update your branch to the latest master.

TonyLianLong committed:
Merge remote-tracking branch 'upstream/master'

TonyLianLong commented:
> It's probably because your patch is not against master but from a point a week or so ago. And in the meantime code size decreased for the minimal port. So you don't really need to worry about it, but to fix it you can update your branch to the latest master.

Okay, then I just wait for them to review.

dpgeorge commented:
This was reworked a bit and merged in 68c28174d0e0ec3f6b1461aea3a0b6a1b84610bb.  To save space the error message itself was removed, so it just raises a plain UnicodeError (which should be enough to warn users what is happening, and is still useful to programatically catch).

The new feature (checking of valid utf8) is controlled by MICROPY_PY_BUILTINS_STR_UNICODE_CHECK and enabled if unicode is enabled.

Thanks @TonyLianLong for the contribution!
